ext/nnstreamer/tensor_filter/tensor_filter_cpp.cc

```diff
@@ -104,37 +166,70 @@ static int cpp_invoke (const GstTensorFilterProperties *prop, void **private_dat
 /**
  * @brief Standard tensor_filter callback
  */
-static int cpp_open (const GstTensorFilterProperties *prop, void **private_data)
+int tensor_filter_cpp_internal::open (const GstTensorFilterProperties *prop, void **private_data)
 {
-  /** @todo: load the class with the given name (either file path or registered name) */
-  /** @todo: configure values of NNS_support_cpp (allocate_in_invoke) */
-  return -EPERM; /** NYI */
+  loadClass(cpp, private_data);
+
+  if (*private_data) {
+    /** Reloading. Unload it to reload */
+    close (prop, private_data);
+    *private_data = NULL;
+  }
+
+  if (filters.find(prop->model_files[0]) == filters.end()) {
+    /* model_files may be really path to .so file. try to dlopen it */
+    if (prop->num_models < 2)
+      return -EINVAL;
+
+    void *handle = dlopen (prop->model_files[1], RTLD_NOW);
+    if (!handle) {
+      return -EINVAL; /** Model file / name not found */
+    }
+
+    /** At this stage, the constructor of the .so file should have registered
+      * itself with tensor_filter_cpp_register() */
+
+    if (filters.find(prop->model_files[0]) == filters.end()) {
+      /** It's still not found. it's not there. */
+      dlclose (handle);
+      g_critical ("tensor_filter_cpp %s is not found in %s.\n",
+          prop->model_files[0], prop->model_files[1]);
+      return -EINVAL;
+    }
+
+    /** Associate handle with the filter. */
+    filters[prop->model_files[0]]->handle = handle;
+
+  }
+
+  *private_data = cpp = filters[prop->model_files[0]];
+  cpp->ref_count++;
+  cpp->prop = prop;
+
+  NNS_support_cpp.allocate_in_invoke = ! cpp->isAllocatedBeforeInvoke();
+
+  return 0;
 }
 
 /**
  * @brief Standard tensor_filter callback
  */
-static void cpp_close (const GstTensorFilterProperties *prop, void **private_data)
+void tensor_filter_cpp_internal::close (const GstTensorFilterProperties *prop, void **private_data)
 {
-  /** @todo Implement This! NYI! */
-  g_assert(false); /** NYI */
-}
-
-static gchar filter_subplugin_cpp[] = "cpp";
+  loadClass(cpp, private_data);
 
-static GstTensorFilterFramework NNS_support_cpp = {
-  .name = filter_subplugin_cpp,
-  .allow_in_place = FALSE,      /** @todo: support this to optimize performance later. */
-  .allocate_in_invoke = FALSE,
-  .run_without_model = FALSE,
-  .verify_model_path = FALSE,
-  .invoke_NN = cpp_invoke,
-  .getInputDimension = cpp_getInputDim,
-  .getOutputDimension = cpp_getOutputDim,
-  .setInputDimension = cpp_setInputDim,
-  .open = cpp_open,
-  .close = cpp_close,
-};
+  /** The class is deallocated from unordered_map if ref_count hits 0 */
+  cpp->ref_count--;
+
+  if (cpp->ref_count == 0) {
+    size_t num = filters.erase(cpp->name);
+    if (num != 1) {
+      g_critical("cpp_close failed to remove %s from unordered map\n",
+          cpp->name);
+    }
+    dlclose (cpp->handle);
```

Comment:
How about checking the validity of `cpp->handle` as well.

Reply:
In the new commit, I'll drop the handle property and close handle right after open.